File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        uniqueChars = collections.defaultdict(int) # char: index
        longest = 0
        l = 0
        for r in range(len(s)):
            if uniqueChars.get(s[r], -1) >= l:
                l = uniqueChars[s[r]] + 1

            uniqueChars[s[r]] = r
            longest = max(longest, r-l+1)

        return longest

        

File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.cpp
class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        unordered_map<char, int> i; // char: index
        int l = 0;
        int longestSubstring = 0;
        for (int r = 0; r < s.size(); r++) {
            if (i.find(s[r]) != i.end() && i[s[r]] >= l) {
                l = i[s[r]] + 1;
            }
            i[s[r]] = r;
            longestSubstring = max(longestSubstring, r-l+1);
        }
        return longestSubstring;

    }
};